Major refactor

NicolasL-S opened the pull request:
This refactor introduces braking changes by changing the output of SpeedMapping to a SpeedMappingResult type.

New features:

- Added a new algorithm, Anderson Acceleration, with recently developed options like composition and adaptive relaxation.
- Made the code faster.
- Added benchmarks comparing different packages with the same functions.
- ACX can now be used with mutable AbstractArrays, StaticArrays, Tuples, and scalars.
- Fixed the documentation.

oscardssmith commented:
Should this get a release?

NicolasL-S commented:
Yes, of course. I'm glad to see there is interest in this old package even if I didn't have time to update it before now. I just created two other repos to provide problems for benchmarking it and I wanted to register them both before I release it. They are both registered now. (I'm not going to register this specific commit, because I made other changes to main after.)

oscardssmith commented:
makes sense. I only care because this package is used in NonlinearSolve's tests and it is holding back ForwardDiff compatibility :)

NicolasL-S commented:
Yes I noticed exactly that when I did my benchmarks against NonlinearSolve. I will do it as soon as possible.

NicolasL-S commented:
Hi @oscardssmith, I just registered SpeedMapping v0.4.0. It can be updated in NonlinearSolve, though note that there was a breaking change: the output is now a struct (with nearly identical fields as the tuple from v0.3.0). Should I open an issue on NonlinearSolve or are they notified automatically of the new version?

oscardssmith commented:
we'll get automatically notified, but if you wanted to make the update pr that would save some effort :)

NicolasL-S commented:
Sure. I agree it would be very helpful for me to update the interface; it's the least I can do. Compared to the working paper and published version, I deprecated some options that were very marginally useful to simplify the interface. But I also added a new algorithm, Anderson acceleration, and new options. I will use my best judgement to suggest a few options to add, but feel free to remove them if necessary. I also changed a few option names and changed the output type. (Now that the semester has started, I'm quite short on time, but I'll try my best to help:)).

For starters, I suggest modifying three files:
- https://github.com/SciML/NonlinearSolve.jl/blob/e4d43ea04e0af7677f0b19a40235b0afa60eeabe/ext/NonlinearSolveSpeedMappingExt.jl (the wrapper)
- https://github.com/SciML/NonlinearSolve.jl/blob/e4d43ea04e0af7677f0b19a40235b0afa60eeabe/src/extension_algs.jl (the interface and docstring for SpeedMapping
- https://github.com/SciML/NonlinearSolve.jl/blob/e4d43ea04e0af7677f0b19a40235b0afa60eeabe/docs/src/refs.bib (updating the reference to the published version and adding a new working paper)

I noticed that SpeedMapping was also used elsewhere in benchmarks and tests with options that may be deprecated. I will try to identify them.

NicolasL-S commented:
I forked NonlinearSolve and created https://github.com/NicolasL-S/NonlinearSolve.jl_updating_SpeedMapping
All the new tests for SpeedMappingJL work (except the already broken tests for the other solvers). The docstring looks good. I was unable to test creating the documentation without the manifest.toml though.

Should I go on and make a pr?

Here are some comments on the changes:

I updated 5 files:
- project.toml
- ext/NonlinearSolveSpeedMappingExt.jl
- src/extension_algs.jl
- test/wrappers/fixedpoint_tests.jl
- docs/src/refs.bib

I made sure that all types in struct `SpeedMappingJL` were visible to the compiler to reduce overhead. For many quantities, a user could supply a Float or an Integer. You could do the same for the other solvers to reduce their overhead too.

I updated `SciMLBase.NLStats(sol.maps, 0, 0, 0, sol.iterations)` with the number of nsolves that the `:aa` algorithm needs (they are done from a QR decomposition). They are still 0 for `:acx`. If this is not what is intended by NLStats, feel free to put back 0.

All the options from v0.3 were either deprecated (`σ_min`, `stabilize`), implemented differently (`check_obj`) or removed because they are not significant enough (`orders`).
Since it is easier to remove than to add, I put all the options from v0.4.0 I thought may be useful. You can read the docstring and decide whether you want to remove some of them, or just reduce the details of the explanations. In my mind, the least important options may be `abstol_obj_grow` and `abstol_obj_grow`. 

I didn't include the possibility to preallocate the caches as this would require importing the relevant types. Anyways, allocation doesn't take too long.

The previous version of `SpeedMappingJL` used the ACX algorithm which is only guaranteed to converge for linear maps or contraction mapping (with an extra assumption) (at least, that's what I managed to show in the paper). In v0.4, I incorporated Anderson acceleration (AA), which works fairly well for any nonlinear problem. So I decided to make regular AA the default option for robustness. Since regular AA is often slower than ACX or AA with `ada_relax` if the mapping is known to converge, users can specify those options manually when appropriate.

In v0.4.0, `:acx` works with scalars, tuples or StaticArrays. But since `:aa` doesn't, I didn't change the requirement to use only mutable arrays for `SpeedMappingJL`.

About testing: 
- I only tested specifications for which the answer is the same as for the other algorithms and that didn't require creating new tests. So no tests for `f`, `lower` and `upper` for instance. But since these options are quite unique to `SpeedMapping`, it makes it complementary to other solvers. So I think users could benefit from them. 
- The default `abstol` is not 1e-10, so I added the option to make sure the tests pass. 
- I tested 4 different specs, you be the judge whether that's too many or too few.
- I commented out the line `B = Hermitian(ones(10) * ones(10)' .* im + Diagonal(1:10))` since `B` is not used.

oscardssmith commented:
that would be great! thanks so much!

NicolasL-S commented:
I think I can't create the PR. I checked the [Github docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork), maybe that explains it:

Note
To open a pull request in a public repository, you must have write access to the head or the source branch or, for organization-owned repositories, you must be a member of the organization that owns the repository to open a pull request.

Could someone else do it or should I become a member?

NicolasL-S commented:
I just realized I forgot to push the change to one line of tests yesterday. Just did it.

NicolasL-S commented:
That last change triggered actions, unlike yesterday. Many errors for reasons way above my head. I think it's probably best you look at my changes in the 5 files I touched on and take it from here.

oscardssmith commented:
will do. Thanks for the help!
